### PR TITLE
transfer skill functionality from user service

### PR DESCRIPTION
**Pull request by grinakate**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/956

### CODE REVIEW

File: src/EducationService.Business/Commands/Skill/CreateSkillCommand.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      _mapper = mapper;
+      _responseCreator = responseCreator;
+    }
+
+    public async Task<OperationResultResponse<Guid?>> ExecuteAsync(CreateSkillRequest request)
+    {
+      if (!await _accessValidator.HasRightsAsync(Rights.AddEditRemoveUsers))
+      {
+        return _responseCreator.CreateFailureResponse<Guid?>(HttpStatusCode.Forbidden);
+      }
```
Comment (MariyaSin) [inline src/EducationService.Business/Commands/Skill/CreateSkillCommand.cs:49]: add empty line

File: src/EducationService.Business/Commands/Skill/CreateSkillCommand.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      if (!validationResult.IsValid)
+      {
+        return _responseCreator.CreateFailureResponse<Guid?>(HttpStatusCode.BadRequest,
+          validationResult.Errors.Select(vf => vf.ErrorMessage).ToList());
+      }
+
+      OperationResultResponse<Guid?> response = new();
+
+      response.Body = await _repository.CreateAsync(_mapper.Map(request));
+      response.Status = OperationResultStatusType.FullSuccess;
```
Comment (MariyaSin) [inline src/EducationService.Business/Commands/Skill/CreateSkillCommand.cs]: the value sets by default

File: src/EducationService.Business/Commands/Skill/CreateSkillCommand.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+      _httpContextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.Created;
+
+      if (response.Body is null)
+      {
+        response = _responseCreator.CreateFailureResponse<Guid?>(HttpStatusCode.BadRequest);
+        response.Status = OperationResultStatusType.Failed;
```
Comment (MariyaSin) [inline src/EducationService.Business/Commands/Skill/CreateSkillCommand.cs]: response creator set the Failed status. No need write it

File: src/EducationService.Data/UserSkillRepository.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    {
+      _provider = provider;
+    }
+
+    public async Task<List<DbUserSkill>> FindAsync(Guid userId)
+    {
+      return (
+        await _provider.UsersSkills
+        .Include(us => us.Skill)
+        .Where(us => us.UserId == userId)
```
Comment (MariyaSin) [inline src/EducationService.Data/UserSkillRepository.cs]: and us.IsActive is true

File: src/EducationService.Mappers/Models/UserSkillDataMapper.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    {
+      if (dbUserSkill is null)
+      {
+        return null;
+      }
+
+      return new UserSkillData(
+        id: dbUserSkill.Id,
+        name: dbUserSkill.Skill.Name
+        );
```
Comment (MariyaSin) [inline src/EducationService.Mappers/Models/UserSkillDataMapper.cs]: move it to 18 line

File: src/EducationService.Models.Db/DbSkill.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    public const string TableName = "Skills";
+
+    public Guid Id { get; set; }
+    public string Name { get; set; }
+    public bool IsActive { get; set; }
+    public Guid CreatedBy { get; set; }
+    public DateTime CreatedAtUtc { get; set; }
+    public Guid? ModifiedBy { get; set; }
+    public DateTime? ModifiedAtUtc { get; set; }
+    public ICollection<DbUserSkill> UserSkills { get; set; }
```
Comment (MariyaSin) [inline src/EducationService.Models.Db/DbSkill.cs]: usersSkill. The skill can have a lot of users

File: src/EducationService.Validation/Skill/CreateSkillRequestValidator.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+using LT.DigitalOffice.EducationService.Validation.Skill.Interfaces;
+
+namespace LT.DigitalOffice.EducationService.Validation.Skill
+{
+  public class CreateSkillRequestValidator : AbstractValidator<CreateSkillRequest>, ICreateSkillRequestValidator
+  {
+    public CreateSkillRequestValidator(ISkillRepository skillRepository)
+    {
+      RuleFor(s => s.Name.Trim())
+          .NotEmpty().WithMessage("Name of Skill must not be empty.")
```
Comment (MariyaSin) [inline src/EducationService.Validation/Skill/CreateSkillRequestValidator.cs]: fix tab

File: src/EducationService.Validation/Skill/CreateSkillRequestValidator.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+using LT.DigitalOffice.EducationService.Models.Dto.Requests.Skills;
+using LT.DigitalOffice.EducationService.Validation.Skill.Interfaces;
+
+namespace LT.DigitalOffice.EducationService.Validation.Skill
+{
+  public class CreateSkillRequestValidator : AbstractValidator<CreateSkillRequest>, ICreateSkillRequestValidator
+  {
+    public CreateSkillRequestValidator(ISkillRepository skillRepository)
+    {
+      RuleFor(s => s.Name.Trim())
```
Comment (MariyaSin) [inline src/EducationService.Validation/Skill/CreateSkillRequestValidator.cs]: you get null ref if you get the null in request

File: src/EducationService.Validation/Skill/CreateSkillRequestValidator.cs [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+using LT.DigitalOffice.EducationService.Models.Dto.Requests.Skills;
+using LT.DigitalOffice.EducationService.Validation.Skill.Interfaces;
+
+namespace LT.DigitalOffice.EducationService.Validation.Skill
+{
+  public class CreateSkillRequestValidator : AbstractValidator<CreateSkillRequest>, ICreateSkillRequestValidator
+  {
+    public CreateSkillRequestValidator(ISkillRepository skillRepository)
+    {
+
```
Comment (MariyaSin) [inline src/EducationService.Validation/Skill/CreateSkillRequestValidator.cs]: remove empty line

**Review by MariyaSin — APPROVED** [review]

File: src/EducationService.Data/SkillRepository.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+      _provider.Skills.Add(skill);
+      await _provider.SaveAsync();
+
+      return skill.Id;
+    }
+
+    public async Task<bool> DoesSkillAlreadyExistAsync(string skillName)
+    {
+      return await _provider.Skills.AnyAsync(s => s.Name == skillName);
```
Comment (VladSpitsin) [inline src/EducationService.Data/SkillRepository.cs]: maybe use ignore-case?

**Review by VladSpitsin — APPROVED** [review]

**Review by ekaterinaryzhenkova — APPROVED** [review]

File: src/EducationService.Data/Interfaces/IUserSkillRepository.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.EducationService.Data.Interfaces
+{
+  [AutoInject]
+  public interface IUserSkillRepository
+  {
+    Task<List<DbUserSkill>> FindAsync(Guid userId);
+  }
+}
```
Comment (irlirla) [inline src/EducationService.Data/Interfaces/IUserSkillRepository.cs]: Add an empty new line at the end of file

File: src/EducationService.Data/Interfaces/ISkillRepository.cs [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+namespace LT.DigitalOffice.EducationService.Data.Interfaces
+{
+  [AutoInject]
+  public interface ISkillRepository
+  {
+    Task<Guid?> CreateAsync(DbSkill skill);
+
+    Task<bool> DoesSkillAlreadyExistAsync(string skillName);
+  }
+}
```
Comment (irlirla) [inline src/EducationService.Data/Interfaces/ISkillRepository.cs]: Add an empty new line at the end of file

File: src/EducationService.Business/Commands/Skill/CreateSkillCommand.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+      if (response.Body is null)
+      {
+        response = _responseCreator.CreateFailureResponse<Guid?>(HttpStatusCode.BadRequest);
+      }
+
+      return response;
+    }
+  }
+}
```
Comment (irlirla) [inline src/EducationService.Business/Commands/Skill/CreateSkillCommand.cs]: Add an empty new line at the end of file

File: src/EducationService.Business/Commands/Skill/Interfaces/ICreateSkillCommand.cs [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.EducationService.Business.Commands.Skill.Interfaces
+{
+  [AutoInject]
+  public interface ICreateSkillCommand
+  {
+    Task<OperationResultResponse<Guid?>> ExecuteAsync(CreateSkillRequest request);
+  }
+}
```
Comment (irlirla) [inline src/EducationService.Business/Commands/Skill/Interfaces/ICreateSkillCommand.cs]: Add an empty new line at the end of file

File: src/EducationService.Validation/Skill/CreateSkillRequestValidator.cs [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+    {
+      RuleFor(s => s.Name)
+        .Cascade(CascadeMode.Stop)
+        .Must(s => string.IsNullOrWhiteSpace(s)).WithMessage("Name of Skill must not be empty.")
+        .Must(s => s.Trim().Length <= 100).WithMessage("Name of Skill is too long.")
+        .MustAsync(async (name, _) => !await skillRepository.DoesSkillAlreadyExistAsync(name))
+        .WithMessage("Skill with this name already exists.");
+    }
+  }
+}
```
Comment (irlirla) [inline src/EducationService.Validation/Skill/CreateSkillRequestValidator.cs]: Add an empty new line at the end of file

File: src/EducationService.Validation/Skill/Interfaces/ICreateSkillRequestValidator.cs [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+using LT.DigitalOffice.EducationService.Models.Dto.Requests.Skills;
+using LT.DigitalOffice.Kernel.Attributes;
+
+namespace LT.DigitalOffice.EducationService.Validation.Skill.Interfaces
+{
+  [AutoInject]
+  public interface ICreateSkillRequestValidator : IValidator<CreateSkillRequest>
+  {
+  }
+}
```
Comment (irlirla) [inline src/EducationService.Validation/Skill/Interfaces/ICreateSkillRequestValidator.cs]: Add an empty new line at the end of file

File: src/EducationService/Controllers/SkillController.cs [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+  {
+    [HttpPost("create")]
+    public async Task<OperationResultResponse<Guid?>> Create(
+     [FromServices] ICreateSkillCommand command,
+     [FromBody] CreateSkillRequest request)
+    {
+      return await command.ExecuteAsync(request);
+    }
+  }
+}
```
Comment (irlirla) [inline src/EducationService/Controllers/SkillController.cs]: Add an empty new line at the end of file

**Review by irlirla — APPROVED** [review]

**Pull request merged**